src/JsonReader.cs

```diff
@@ -548,127 +570,162 @@ sealed class NonProperty : IJsonProperty<Unit, JsonReadResult<Unit>>
-                else if (!reader.TrySkip())
-                    throw PartialJsonNotSupportedException();
+                            DefaultUnassigned(property1,  ref values.V1);
+                            DefaultUnassigned(property2,  ref values.V2);
+                            DefaultUnassigned(property3,  ref values.V3);
+                            DefaultUnassigned(property4,  ref values.V4);
+                            DefaultUnassigned(property5,  ref values.V5);
+                            DefaultUnassigned(property6,  ref values.V6);
+                            DefaultUnassigned(property7,  ref values.V7);
+                            DefaultUnassigned(property8,  ref values.V8);
+                            DefaultUnassigned(property9,  ref values.V9);
+                            DefaultUnassigned(property10, ref values.V10);
+                            DefaultUnassigned(property11, ref values.V11);
+                            DefaultUnassigned(property12, ref values.V12);
+                            DefaultUnassigned(property13, ref values.V13);
+                            DefaultUnassigned(property14, ref values.V14);
+                            DefaultUnassigned(property15, ref values.V15);
+                            DefaultUnassigned(property16, ref values.V16);
+
+                            return (values.V1, values.V2, values.V3,
+                                    values.V4, values.V5, values.V6,
+                                    values.V7, values.V8, values.V9,
+                                    values.V10, values.V11, values.V12,
+                                    values.V13, values.V14, values.V15,
+                                    values.V16) is ((true, var v1), (true, var v2), (true, var v3),
+                                                    (true, var v4), (true, var v5), (true, var v6),
+                                                    (true, var v7), (true, var v8), (true, var v9),
+                                                    (true, var v10), (true, var v11), (true, var v12),
+                                                    (true, var v13), (true, var v14), (true, var v15),
+                                                    (true, var v16))
+                                 ? Value(projector(v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11, v12, v13, v14, v15, v16))
+                                 : Error("Invalid JSON object.");
+
+                        case ObjectReadStateMachine.ReadResult.PropertyName:
+                            static bool TrySetPropertyIndex<TValue>(int index,
+                                                                    IJsonProperty<TValue, JsonReadResult<TValue>> property,
+                                                                    ref Utf8JsonReader reader,
+                                                                    ref ObjectValueState<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16> values)
```

Comment:
Instead of passing the whole `ObjectValueState<…>` by reference, you could just pass a reference to the index, as in:

```c#
static bool TrySetPropertyIndex<TValue>(int index,
                                        IJsonProperty<TValue, JsonReadResult<TValue>> property,
                                        ref Utf8JsonReader reader,
                                        ref int? currentIndex)
{
    if (!property.IsMatch(ref reader))
        return false;

    currentIndex = index;
    return true;
}

ref var currentPropertyIndex = ref values.CurrentPropertyIndex;

_ =    TrySetPropertyIndex(1,  property1,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(2,  property2,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(3,  property3,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(4,  property4,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(5,  property5,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(6,  property6,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(7,  property7,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(8,  property8,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(9,  property9,  ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(10, property10, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(11, property11, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(12, property12, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(13, property13, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(14, property14, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(15, property15, ref reader, ref currentPropertyIndex)
    || TrySetPropertyIndex(16, property16, ref reader, ref currentPropertyIndex);
```

Reply:
Applied with c2099a95eb80da5ff4053f931fad1fefd0eae74c.